text/projects/Szymon_Salabura/Teng-Yao-SCS.tex

```diff
@@ -0,0 +1,188 @@
+\documentclass[12pt]{article}
+\usepackage{polski}
+\usepackage[utf8]{inputenc}
+\usepackage{geometry}
+\usepackage{mathtools}
+\usepackage{amsthm}
+\usepackage{hyperref}
+\newgeometry{left=1in,right=1in,top=1in,bottom=1in}
+
+\newtheorem{theorem}{Twierdzenie}[section]
+\newtheorem{lemma}{Lemat}[section]
+
+\begin{document}
+
+\begin{center}
+    \Large{Algorytm 2.89-aproksymacyjny \\ dla problemu najkrótszego wspólnego nadsłowa} \\
+    \large{autorstwa Shang-Hua Teng oraz Frances F. Yao \cite{Teng-Yao}}
+\end{center}
+
+\section{Notacja i definicje}
+
+Dla danego na wejściu zbioru słów $S$ chcemy znaleźć najkrótsze słowo $\alpha$ będące nadsłowem tego zbioru, t.j. $\forall_{s \in S} \, s$ jest podsłowem $\alpha$.
+Standardowo, zakładamy że dla żadnej pary słów nie zachodzi sytuacja, w której jedno jest podsłowem drugiego.
+Wprowadzamy następujące definicje i oznaczenia:
+\begin{itemize}
+\item $|s|$ jest długością słowa $s$,
+\item $|S|=\sum_{s \in S}\,|s|$,
+\item $opt(S)$ jest długością wyniku optymalnego - najkrótszego nadsłowa słów ze zbioru $S$.
+\end{itemize}
+Dla każdej pary słów $s=uv$, $t=vw$:
+\begin{itemize}
+\item $ov(s,t)=|v|$ jest długością ``nakładania się'' na siebie słów $s$ i $t$,
+\item $pref(s,t)=u$.
+\end{itemize}
+Dla takich słów $s$ i $t$ najkrótszym ich nadsłowem, w którym $s$ występuje przed $t$ jest $uvw=pref(s,t)t$.
+Słowo takie nazywamy złączeniem słów $s$ i $t$ oraz oznaczamy je przez $\langle s,t \rangle$.
+Łatwo zauważyć, że najkrótsze nadsłowo $\alpha$ jest równe $\langle s_{\pi(1)},s_{\pi(2)},\dots,s_{\pi(n)} \rangle$ dla pewnej permutacji $\pi$ zbioru $\{1,2,\dots,n\}$. \\
+
+Problem znajdowania najkrótszego nadsłowa może zostać przeformułowany na problem grafowy.
+Dla zbioru słów $S$ możemy stworzyć ważony graf skierowany $G=(S,E)$, w którym wagi krawędzi są równe $d(s,t)=|pref(s,t)|$.
+Dla zbioru krawędzi zdefiniujmy $d(E)=\sum_{(s,t) \in E}\,d(s,t)$ i analogicznie $ov(E)=\sum_{(s,t) \in E}\,ov(s,t)$.
+Prosta obserwacja pokazuje, że suma długości słów w cyklu $E$ jest równa $d(E)+ov(E)$. 
+
+\section{Optymalne pokrycie cyklowe}
+
+Pokryciem cyklowym grafu $G$ jest zbiór cykli taki, że każdy wierzchołek $G$ należy do któregoś z nich.
+Optymalne pokrycie cyklowe jest pokryciem cyklowym z (w tym przypadku) najmniejszą możliwą sumą wag krawędzi.
+Można je znaleźć w czasie $O(n^3)$ algorytmem węgierskim \footnote{\url{https://en.wikipedia.org/wiki/Hungarian_algorithm}}. \\
```

Comment:
Shouldn't any max flow algorithm work as well?

Reply:
Yes, but I wanted to focus on the Hungarian Algorithm, because that's what I implemented.
Added 'między innymi'.